### PR TITLE
Out of order ledger catchup recovery

### PR DESCRIPTION
**Pull request by sisuresh**
# Description

This PR includes Marta's `Herder` change that will allow out of order ledgers into `LedgerManager`.  `CatchupManager` was changed to account for out of order ledgers, and potentially recover if the missing ledgers are received before `CatchupWork` starts.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/catchup/CatchupManagerImpl.cpp [hunk 1]
```diff
@@ -79,59 +89,65 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
...
-        hm.ledgerToTriggerCatchup(firstLedgerInBuffer);
-    if (lastReceivedLedgerSeq >= catchupTriggerLedger)
+    auto& hm = mApp.getHistoryManager();
+
+    uint32_t lastLedgerInBuffer = mSyncingLedgers.crbegin()->first;
+    uint32_t firstLedgerInCheckpoint =
+        hm.firstLedgerInCheckpointContaining(lastLedgerInBuffer);
+
+    auto it = mSyncingLedgers.find(firstLedgerInCheckpoint);
+    if (it != mSyncingLedgers.end() &&
```
Comment (sisuresh) [inline src/catchup/CatchupManagerImpl.cpp:124]: @MonsieurNicolas I was thinking about ways to make this condition simpler, but I think looking for the start of the checkpoint based on the largest buffered ledger is the best way to do this. I could also address https://github.com/stellar/stellar-core/issues/2455 if you think it's worth doing now, which will change this code a bit.

File: src/ledger/LedgerManagerImpl.cpp [hunk 2]
```diff
@@ -523,6 +528,15 @@ during replays.
 void
 LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
 {
+    if (mLastClosedLedger.hash != ledgerData.getTxSet()->previousLedgerHash())
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/ledger/LedgerManagerImpl.cpp]: Good catch. I removed this.

File: src/ledger/LedgerManagerImpl.h [hunk 3]
```diff
@@ -75,7 +75,8 @@ class LedgerManagerImpl : public LedgerManager
     {
         CLOSED,
         TOO_OLD,
-        TOO_NEW
+        TOO_NEW,
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/ledger/LedgerManagerImpl.h]: Good idea. I removed the enum and moved the `setState` calls into `closeLedgerIf`.

File: src/catchup/CatchupManagerImpl.cpp [hunk 4]
```diff
@@ -79,59 +89,65 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
...
+    // buffered ledger
+
     std::string message;
-    uint32_t catchupTriggerLedger =
-        hm.ledgerToTriggerCatchup(firstLedgerInBuffer);
-    if (lastReceivedLedgerSeq >= catchupTriggerLedger)
+    auto& hm = mApp.getHistoryManager();
+
+    uint32_t lastLedgerInBuffer = mSyncingLedgers.crbegin()->first;
+    uint32_t firstLedgerInCheckpoint =
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp]: Done. I added a variable to track the smallest buffered checkpoint ledger. The last commit has the change.

File: src/catchup/CatchupManagerImpl.cpp [hunk 5]
```diff
@@ -57,20 +56,31 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
...
-    // If LCL is already at-or-ahead of the ledger we just received from the
-    // network, we're up to date. Return early, nothing to do.
-    if (lastReceivedLedgerSeq <=
-        mApp.getLedgerManager().getLastClosedLedgerNum())
+    // 1. CatchupWork is not running yet
+    // 2. We've seen out of sync ledgers
+    // 3. The ledger just received is equal to lcl
+    // then it's possible we're back in sync and we can attempt to apply
+    // mSyncingLedgers
+    if (!mCatchupWork && !mSyncingLedgers.empty() &&
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp:64]: If the condition is moved down, the `lastReceivedLedgerSeq <= mApp.getLedgerManager().getLastClosedLedgerNum()` condition (line 71) will also need to be changed to allow ledgers equivalent to lcl to flow through. I think this is an issue because we could potentially buffer a ledger equal to lcl while catchup is running. The `ledgerData.getLedgerSeq() <= config.toLedger()` condition (line 96) wouldn't help in a scenario where we are in the middle of applying buffered ledgers, which would take lcl past `config.toLedger()`. If a ledger equal to lcl came in at this point, we would buffer it.

This scenario sounds impossible (you'd have to externalize the same ledger twice), but I'm not sure if that assumption can be made. We could just duplicate the lcl check if catchup is running, but at that point I think what we have here is cleaner.

File: src/ledger/LedgerManagerImpl.cpp [hunk 6]
```diff
@@ -427,55 +427,39 @@ LedgerManagerImpl::valueExternalized(LedgerCloseData const& ledgerData)
...
+                << "Can't close ledger: " << ledgerAbbrev(mLastClosedLedger)
+                << "  in LM because catchup is running";
+            return;
         }
+
+        closeLedger(ledgerData);
+        CLOG(INFO, "Ledger")
+            << "Closed ledger: " << ledgerAbbrev(mLastClosedLedger);
+
+        setState(LM_SYNCED_STATE);
```
Comment (marta-lokhova) [inline src/ledger/LedgerManagerImpl.cpp:455]: I think it makes sense to mark ledger manager as "synced" when the last closed ledger is the newest ledger we know of. This is not the case when LM has seen some future ledger (at which point CatchupManager has started buffering ledgers). So I think when catchup has buffered some newer ledgers, LM should stay in "catching up" state.

Comment (sisuresh) [reply src/ledger/LedgerManagerImpl.cpp]: Done. I ended up keeping the requirement to close the next ledger before moving to the synced state so the logic stays consistent with how catchup works.

File: src/herder/test/HerderTests.cpp [hunk 7]
```diff
@@ -1890,7 +1890,7 @@ TEST_CASE("quick restart", "[herder][quickRestart]")
     auto static const SMALL_GAP = app->getConfig().MAX_SLOTS_TO_REMEMBER - 1;
     // BIG_GAP, we just need to pick a number greater than what we keep in
     // memory
-    auto static const BIG_GAP = app->getConfig().MAX_SLOTS_TO_REMEMBER + 1;
+    auto static const BIG_GAP = app->getConfig().MAX_SLOTS_TO_REMEMBER + 2;
```
Comment (sisuresh) [inline src/herder/test/HerderTests.cpp:1893]: @marta-lokhova I had to increase this for the test to pass.

File: src/catchup/CatchupManagerImpl.cpp [hunk 8]
```diff
@@ -79,59 +88,78 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
...
         return;
     }
 
-    // Next switch between 3 cases: starting buffering, waiting to
-    // buffer, and already buffering.
+    // Next, we buffer every out of sync ledger to allow us to get back in sync
+    // in case the ledgers we're missing are received.
+    addToSyncingLedgers(ledgerData);
+
+    // Keep track of the smallest buffered checkpoint.
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupManagerImpl.cpp:108]: I don't see the point in dealing with the complexity of managing this member `mSmallestBufferedCheckpoint `: you can easily compute it given the map of buffered ledgers (just like how you compute the "highest checkpoint" in `trimSyncingLedgers`).

File: src/catchup/CatchupManagerImpl.cpp [hunk 9]
```diff
@@ -79,59 +88,78 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
...
+        // will set this again if a checkpoint still exists in mSyncingLedgers
+        mSmallestBufferedCheckpoint.reset();
     }
     else
     {
-        auto eta = (catchupTriggerLedger - lastReceivedLedgerSeq) *
+        uint32_t requiredFirstLedgerInCheckpoint =
+            mSmallestBufferedCheckpoint
+                ? *mSmallestBufferedCheckpoint
+                : hm.firstLedgerAfterCheckpointContaining(lastLedgerInBuffer);
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupManagerImpl.cpp:153]: I don't understand this: why do we use the checkpoint derived from the last ledger in buffer, shouldn't it be the checkpoint derived from the first ledger instead as this is what we're going to use?

Comment (MonsieurNicolas) [reply src/catchup/CatchupManagerImpl.cpp:153]: if we have gaps where we're missing the checkpoint ledger, it's probably fine to give the ETA assuming we're going to recover from that gap (so aim to get the smallest checkpoint). I guess for this particular scenario, the ETA should be 0 as we should be getting that ledger "any time".

File: src/history/test/HistoryTests.cpp [hunk 10]
```diff
@@ -1264,3 +1264,142 @@ TEST_CASE("Catchup failure recovery with buffered checkpoint",
...
+    auto app = catchupSimulation.createCatchupApplication(
+        std::numeric_limits<uint32_t>::max(), Config::TESTDB_IN_MEMORY_SQLITE,
+        "app2");
+
+    auto checkpointLedger = catchupSimulation.getLastCheckpointLedger(1);
+    catchupSimulation.ensureOnlineCatchupPossible(checkpointLedger, 15);
+
+    // randomize externalize order of buffered ledgers
+    REQUIRE(
+        catchupSimulation.catchupOnline(app, checkpointLedger, 10, 0, 0, true));
```
Comment (MonsieurNicolas) [inline src/history/test/HistoryTests.cpp:1282]: I think that instead of a randomize flag it would be better to pass in the actual order of ledgers in an optional vector (it also replaces the `ledgersToInject` vector) : that way you can easily test all the scenarios in this file with this parameter without introducing randomization (that actually doesn't really buy you anything here).
I'm thinking: generate range 1..128 (or whatever you need) and mutate it as you see it fit (for example, you can swap 64 and 65), you can add gaps, whatever you want there

File: src/herder/test/HerderTests.cpp [hunk 11]
```diff
@@ -1617,6 +1617,206 @@ TEST_CASE("SCP State", "[herder][acceptance]")
...
+
+    SCPQuorumSet qset;
+    qset.threshold = 1;
+    qset.validators.push_back(validatorKey.getPublicKey());
+
+    auto cfg1 = getTestConfig(1);
+    auto cfg2 = getTestConfig(2);
+    cfg1.MAX_SLOTS_TO_REMEMBER = 5;
+    // Setup a fake archive to prevent HistoryManager from throwing
+    cfg2.HISTORY["test"] = HistoryArchiveConfiguration{"test"};
```
Comment (MonsieurNicolas) [inline src/herder/test/HerderTests.cpp:1637]: why do we need this fake archive? I don't think we set this in other simulations

File: src/herder/test/HerderTests.cpp [hunk 12]
```diff
@@ -1617,6 +1617,206 @@ TEST_CASE("SCP State", "[herder][acceptance]")
...
+
+        // Ensure Herder still tracks the right index despite receiving older
+        // message
+        REQUIRE(herder.getCurrentLedgerSeq() ==
+                futureSlot.first.statement.slotIndex);
+
+        // Ledger manager is still out of sync, as we have seen future ledgers
+        REQUIRE(!lm.isSynced());
+
+        // complete - all messages are received out of order (10, 7, 9, 8)
```
Comment (MonsieurNicolas) [inline src/herder/test/HerderTests.cpp:1759]: should we just pass it the actual vector of indexes to apply instead of a bool? This is pretty hard to follow. I'm thinking you'd pass in `{10,7, 9,8}` and `{10, 7, 8, 9}`

File: src/herder/test/HerderTests.cpp (hunk 7 above)
Comment (MonsieurNicolas) [reply src/herder/test/HerderTests.cpp:1893]: weird, I don't see why this would be the case. there are no out of order ledgers. @marta-lokhova any idea?

File: src/history/test/HistoryTestsUtils.cpp [hunk 13]
```diff
@@ -766,9 +758,42 @@ CatchupSimulation::catchupOnline(Application::pointer app, uint32_t initLedger,
...
+    // after it
+    if (randomizeBufferedLedgers)
+    {
+        for (uint32_t n = initLedger; n < triggerLedger; ++n)
+        {
+            externalize(n);
+        }
+
+        std::vector<uint32_t> bufferedLedgers(bufferLedgers);
+        std::iota(bufferedLedgers.begin(), bufferedLedgers.end(),
```
Comment (MonsieurNicolas) [inline src/history/test/HistoryTestsUtils.cpp:772]: `std::iota` doesn't build for me (missing include?)

File: src/ledger/LedgerManagerImpl.cpp [hunk 14]
```diff
@@ -452,22 +452,10 @@ LedgerManagerImpl::closeLedgerIf(LedgerCloseData const& ledgerData)
...
-        }
-        else
-        {
-            CLOG(FATAL, "Ledger") << "Network consensus for ledger "
-                                  << mLastClosedLedger.header.ledgerSeq
-                                  << " changed; this should never happen";
-            CLOG(FATAL, "Ledger") << POSSIBLY_CORRUPTED_QUORUM_SET;
-            throw std::runtime_error("Network consensus inconsistency");
-        }
+        closeLedger(ledgerData);
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/ledger/LedgerManagerImpl.cpp:451]: It wasn't causing problems. I ended up removing it because the same check is being done in `closeLedger`.

File: src/catchup/CatchupManagerImpl.cpp (hunk 8 above)
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp:108]: Keeping track of the smallest checkpoint was done because it was more efficient than looking through the buffered ledgers, but the buffered ledgers should never grow that large before catchup starts, so a search should be fine. I'll remove `mSmallestBufferedCheckpoint`.

File: src/catchup/CatchupManagerImpl.cpp (hunk 9 above)
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp:153]: I took the approach of always looking forward, but being optimistic and expecting the checkpoint ledger makes sense since we should be getting it.

File: src/herder/HerderSCPDriver.cpp [hunk 15]
```diff
@@ -792,6 +780,21 @@ HerderSCPDriver::valueExternalized(uint64_t slotIndex, Value const& value)
...
+    //  * when getting back in sync (a gap potentially opened)
+    // in both cases it's safe to not proceed with the rest of the function
+    // as we're already tracking a more recent state
+    // Yet, still deliver externalize events to LedgerManager so it
+    // can process catchup state, if needed
+    if (slotIndex <= mApp.getHerder().getCurrentLedgerSeq())
+    {
+        TxSetFramePtr externalizedSet = mPendingEnvelopes.getTxSet(b.txSetHash);
+        LedgerCloseData ledgerData(slotIndex, externalizedSet, b);
+        mLedgerManager.valueExternalized(ledgerData);
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderSCPDriver.cpp:794]: @MonsieurNicolas I can update it. My original approach was to make herder behave in the same way as master, but maybe a better way would be to split events in Herder into "should happen for every ledger", and "should happen only for future ledgers". Also, do we care about metrics for old ledgers? (recording externalize events etc)

Regarding the second comment, I'm not sure what is the benefit of triggering the next ledger, when we still have some future ledgers buffered? (LedgerManager would technically still be "out of sync", right?)

[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp:794]: yes @marta-lokhova that's what I'd prefer to do: separate what we want to do for any ledger vs most recent.
I don't think tracking metrics for those out of order ledgers is that important so the list of what to do with older ledgers is fairly limited.

second issue (not for this PR) is that we don't even trigger when we close the most recent ledger, we have to close the next ledger (so we're delayed by one ledger).

File: src/herder/test/HerderTests.cpp (hunk 7 above)
Comment (marta-lokhova) [reply src/herder/test/HerderTests.cpp:1893]: hah didn't realize the test is failing; this is related to cfg1, cfg2 fix on line 1841-1842. turns out there is an [off-by-one error](https://github.com/stellar/stellar-core/blob/master/src/herder/HerderImpl.cpp#L234) in ledger validity bracket implementation: we remember `MAX_SLOTS_TO_REMEMBER + 1`  This test was not failing before because listener was 1 ledger behind already before validator started "creating the gap", so it'd fail. So we should fix the math in Herder, and leave "big gap" as is.

File: src/history/test/HistoryTestsUtils.cpp (hunk 13 above)
Comment (sisuresh) [reply src/history/test/HistoryTestsUtils.cpp]: removed iota

File: src/history/test/HistoryTests.cpp [hunk 16]
```diff
@@ -1309,19 +1309,24 @@ TEST_CASE("Introduce and fix gap without starting catchup",
     REQUIRE(!lm.isSynced());
     REQUIRE(cm.hasBufferedLedger());
 
-    // Fill in gap. This will put ledger manager back in sync, but we'll still
-    // have buffered ledgers because there's another gap
+    // Fill in gap. We'll still be out of sync
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/history/test/HistoryTests.cpp]: Fixed comment

File: src/catchup/CatchupManagerImpl.cpp [hunk 17]
```diff
@@ -315,9 +333,32 @@ CatchupManagerImpl::trimSyncingLedgers()
...
-    else
+}
+
+void
+CatchupManagerImpl::tryApplySyncingLedgers()
+{
+    auto const& ledgerHeader =
+        mApp.getLedgerManager().getLastClosedLedgerHeader();
+
+    auto it = mSyncingLedgers.cbegin();
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp:347]: done

File: src/catchup/CatchupManagerImpl.cpp [hunk 18]
```diff
@@ -315,9 +333,32 @@ CatchupManagerImpl::trimSyncingLedgers()
...
+
+        // we still have a missing ledger
+        if (ledgerHeader.header.ledgerSeq + 1 != lcd.getLedgerSeq())
+        {
+            break;
+        }
+
+        mApp.getLedgerManager().closeLedger(lcd);
+        CLOG(INFO, "History")
+            << "Closed buffered ledger: " << ledgerHeader.header.ledgerSeq;
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp]: fixed

File: src/catchup/CatchupManagerImpl.cpp (hunk 8 above)
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.cpp]: Added a `findFirstCheckpoint` method to use here and in `trimSyncingLedgers`.

File: src/history/test/HistoryTests.cpp (hunk 10 above)
Comment (sisuresh) [reply src/history/test/HistoryTests.cpp]: Done

File: src/herder/HerderSCPDriver.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderSCPDriver.cpp]: ok @MonsieurNicolas I made the changes. A few things here: 
* Saving SCP state: when an older slot is received, its envelopes are saved. It looks like we track the most recent transitive quorum, so I skipped saving anything quorum-related for older slots (as we already track newer state, so it seems like the right thing to do?).
* to be more conservative, we currently only allow older slots when herder is in sync (I think that makes sense, since catchup can't start anyway?)
* I'm looking into ways we could be more defensive about this: I notice that a lot of functions make assumptions about last known consensus index, and wondering if there is some invariance we could add now that we're letting older slots to be processed.

File: src/herder/test/HerderTests.cpp (hunk 11 above)
Comment (marta-lokhova) [reply src/herder/test/HerderTests.cpp]: You're right, we do not actually need it (catchup is never started, so no need to set up the archive).

File: src/herder/HerderSCPDriver.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderSCPDriver.cpp]: @MonsieurNicolas re: second issue 
I think this happens in master currently as well. LedgerManager does not go "in sync" right after catchup applied all buffered ledgers - it waits for the next ledger. Since LM is not in sync, Herder skips triggering next ledger.

File: src/herder/HerderSCPDriver.cpp [hunk 19]
```diff
@@ -792,44 +780,74 @@ HerderSCPDriver::valueExternalized(uint64_t slotIndex, Value const& value)
...
+    //  * when getting back in sync (a gap potentially opened)
+    // in both cases do limited processing on older slots; more importantly,
+    // deliver externalize events to LedgerManager
+    bool newSlot = true;
+    auto currentSlot = mApp.getHerder().getCurrentLedgerSeq();
+    // Allow older messages when in sync
+    if (slotIndex < currentSlot && trackingSCP())
     {
-        logQuorumInformation(slotIndex - 2);
+        newSlot = false;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp:793]: maybe call this `isLatestSlot` instead of `newSlot`; also just set it to its value directly instead of burying in conditions:
`isLatestSlot = (slotIndex < mApp.getHerder().getCurrentLedgerSeq());`

File: src/herder/HerderSCPDriver.cpp [hunk 20]
```diff
@@ -792,44 +780,74 @@ HerderSCPDriver::valueExternalized(uint64_t slotIndex, Value const& value)
...
+
+        mTrackingSCP = std::make_unique<ConsensusData>(slotIndex, b);
+
+        if (!mLastTrackingSCP)
+        {
+            mLastTrackingSCP = std::make_unique<ConsensusData>(*mTrackingSCP);
+        }
     }
 
+    auto recordExternalize = [&](bool forceSelf) {
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp:835]: simpler would have been to just put the old code in the `if (newSlot)` block and add an else block that just calls `mHerder.valueExternalized(slotIndex, b);`

File: src/herder/HerderImpl.cpp [hunk 21]
```diff
@@ -148,19 +148,37 @@ void
 HerderImpl::valueExternalized(uint64 slotIndex, StellarValue const& value)
 {
     const int DUMP_SCP_TIMEOUT_SECONDS = 20;
+    bool newSlot = slotIndex == getCurrentLedgerSeq();
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:151]: why this is not
* the same value than what is being computed in SCPDriver?
* a parameter (named "isLatestSlot"?)

File: src/herder/HerderImpl.cpp [hunk 22]
```diff
@@ -148,19 +148,37 @@ void
...
 
-    // record metrics
-    getHerderSCPDriver().recordSCPExecutionMetrics(slotIndex);
+    if (newSlot)
+    {
+        getHerderSCPDriver().recordSCPExecutionMetrics(slotIndex);
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:155]: should this call to `HerderSCPDriver` be moved into `HerderSCPDriver::valueExternalized` ?

File: src/herder/HerderImpl.cpp [hunk 23]
```diff
@@ -180,10 +198,15 @@ HerderImpl::valueExternalized(uint64 slotIndex, StellarValue const& value)
     // save the SCP messages in the database
     if (mApp.getConfig().MODE_STORES_HISTORY)
     {
+        auto qmap = nullopt<QuorumTracker::QuorumMap>();
+        if (newSlot)
+        {
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:203]: why are you trying to not save qmap?

File: src/herder/HerderPersistenceImpl.cpp [hunk 24]
```diff
@@ -93,6 +93,13 @@ HerderPersistenceImpl::saveSCPHistory(uint32_t seq,
         }
     }
 
+    if (!qmapOptional)
```
Comment (MonsieurNicolas) [inline src/herder/HerderPersistenceImpl.cpp:96]: this is wrong: not only you're skipping saving the qmap (what you were trying to do here - but that we should not do), but you're also not saving qset information from the messages that are included in the SCP dump

File: src/herder/HerderImpl.cpp [hunk 25]
```diff
@@ -148,19 +148,37 @@ void
...
+        if (gap > DUMP_SCP_TIMEOUT_SECONDS)
+        {
+            auto slotInfo = getJsonQuorumInfo(getSCP().getLocalNodeID(), false,
+                                              false, slotIndex);
+            Json::FastWriter fw;
+            CLOG(WARNING, "Herder")
+                << fmt::format("Ledger took {} seconds, SCP information:{}",
+                               gap, fw.write(slotInfo));
+        }
+    }
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:169]: Wouldn't it be easier to refactor this code instead of adding a bunch of `if(newSlot)`?

Right now you're still including certain things that should not be included when processing an old slot:
* calls to `trackingHeartBeat()`
* `mTriggerTimer.cancel()`
* `checkAndMaybeReanalyzeQuorumMap()`
* `ledgerClosed()`

basically you're left with only the "central part" of this function that does:
* logging
* saving of SCP message
* cleanup upgrades
* call ledger manager's `valueExternalized`

File: src/catchup/CatchupManagerImpl.h [hunk 26]
```diff
@@ -29,17 +29,24 @@ class CatchupManagerImpl : public CatchupManager
...
     void addToSyncingLedgers(LedgerCloseData const& ledgerData);
     void startOnlineCatchup();
     void trimSyncingLedgers();
     void trimAndReset();
     void tryApplySyncingLedgers();
     uint32_t getCatchupCount();
 
+    template <typename T>
+    T
+    findFirstCheckpoint(T begin, T end)
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/catchup/CatchupManagerImpl.h]: Done. I made `findFirstCheckpoint` a free function and moved it into the `.cpp`

File: src/herder/HerderSCPDriver.cpp (hunk 20 above)
Comment (marta-lokhova) [reply src/herder/HerderSCPDriver.cpp:835]: I'm not sure how this is simpler: what you are proposing involves calling `valueExternalized` in two places; so now if anything about the function signature or its arguments changes, we'd need to modify it in two places, which makes it slightly more error-prone.

File: src/herder/HerderImpl.cpp (hunk 21 above)
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:151]: SCPDriver always updates tracking value for new slots. From what I see, HerderImpl::valueExternalized is only called when Herder is back in sync (and this is enforced by trackingSCP in trackingHeartbeat). If you think it is better to just rely on a parameter, it can be changed (isLatestSlot parameter was my initial implementation anyway; but I found it easier to reason about tracking values when in sync or out of sync, so I'd like to understand why using a parameter would be better here)

File: src/herder/HerderPersistenceImpl.cpp (hunk 24 above)
Comment (marta-lokhova) [reply src/herder/HerderPersistenceImpl.cpp:96]: Discussed offline -- right now persistence logic assumes that we always save the latest state, which might not be the case anymore. The solution is to make changes to sql queries to make sure we're not overriding newer info.

File: src/herder/HerderSCPDriver.cpp (hunk 20 above)
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp:835]: it's simpler: it's trading a copy/paste of 1 line visible within 5 lines of code vs a lambda and a flow that is not super straightforward to understand.

File: src/herder/HerderImpl.cpp (hunk 21 above)
Comment (MonsieurNicolas) [reply src/herder/HerderImpl.cpp:151]: Reason I thought it was a parameter was that `HerderSCPDriver::valueExternalized` seems to be the one defining the criteria for considering processing that slot as "latest" or not.

I guess we have a hidden ordering dependency already so we can leave it like that.

**Comment by MonsieurNicolas** [comment]
r+ a7798d6a8acadcafc98527dc50bbe2772940b3f5
